lua/hop/hint.lua

```diff
@@ -186,13 +225,38 @@ function M.create_hints(hint_mode, win_width, cursor_pos, col_offset, top_line,
   local indirect_hints = {}
   local hint_counts = 0
   for i = 1, #lines do
-    local line_hints = M.mark_hints_line(hint_mode, top_line + i - 1, lines[i], col_offset, win_width)
-    hints[i] = line_hints
 
-    hint_counts = hint_counts + #line_hints.hints
+    local line_nr = top_line + i - 1
+
+    -- A line is too early if it's before the cursor and the motion only
+    -- operates on lines after the cursor.
+    local line_too_early = hint_mode.direction == 'after_cursor' and line_nr + 1 < cursor_pos[1]
+    -- A line is too late if it's after the cursor and the motion only
+    -- operates on lines before the cursor.
+    local line_too_late = hint_mode.direction == 'before_cursor' and line_nr + 1 > cursor_pos[1]
+    -- A line is invalid if it's not on the same line as the cursor and the
+    -- motion only operates on the current line.
+    local line_not_exact = line_nr + 1 ~= cursor_pos[1] and hint_mode.same_line
+    -- A oneshot motion moves between lines so there's no reason to include the
+    -- current line in the motion.
+    local oneshot_excludes_current = line_nr + 1 == cursor_pos[1] and hint_mode.oneshot
+    -- If any of the above bools are true, the line is invalid and shouldn't be hinted.
+    local is_invalid = line_too_early or line_too_late or line_not_exact or oneshot_excludes_current
+
+    if not is_invalid then
+      local line_hints = M.mark_hints_line(hint_mode, line_nr, lines[i], col_offset, win_width, cursor_pos)
+      hints[i] = line_hints
+
+      hint_counts = hint_counts + #line_hints.hints
+    else
+      hints[i] = {
+        hints = {};
+        length = 0
+      }
```

Comment:
Yeah, see, this is why I don’t like the design you suggest. You modified the already existing abstractions and implementations to bend them to do what they were not meant to be used for.  This logic will be applied to lots of lines for nothing.